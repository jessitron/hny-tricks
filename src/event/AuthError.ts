import { html } from "../htm-but-right";
import { currentTraceId } from "../tracing-util";

export class HnyTricksAuthError extends Error {
  constructor(public message: string, public contextMessage: string) {
    super(message);
    this.name = "HnyTricksAuthError";
  }
}

export function isHnyTricksAuthError(
  error: Error
): error is HnyTricksAuthError {
  return error.name === "HnyTricksAuthError";
}

export function HnyTricksAuthErrorMessage(params: {
  error: HnyTricksAuthError;
}) {
  const { error } = params;
  return html`
    <div traceId=${currentTraceId()}>
      <p>Problem with Authorization.</p>
      <p>Message: ${error.message}</p>
      <p>Context: ${error.contextMessage}</p>
    </div>
  `;
}
