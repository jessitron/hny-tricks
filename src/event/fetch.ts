import { trace } from "@opentelemetry/api";
import { recordError } from "../tracing-util";

import JSON5 from "json5";

type SomeResponse = object;

const RECORD_BODY = true; // don't put sensitive data in your weird test event

export type FetchEventError = {
  fetchEventError: true;
  statusCode: number;
  message: string;
};
export function isFetchEventError(
  response: SomeResponse | FetchEventError
): response is FetchEventError {
  return (response as FetchEventError).fetchEventError;
}

export async function fetchEventJson(
  url: string
): Promise<object | object[] | FetchEventError> {
  return fetch(url, {
    method: "GET",
  }).then(
    async (result) => {
      if (!result.ok) {
        recordError("Result not OK from fetching event", {
          "response.statusText": result.statusText,
          "response.status": result.status,
          "response.body": await result.text(),
          "http.url": url,
        });
        return {
          fetchEventError: true,
          statusCode: result.status,
          message: `Received ${result.status} from ${url}. ${result.statusText}`,
        };
      }
      // Do not check the content type; a gist returns us text/plain

      const resultText = await result.text();

      if (RECORD_BODY) {
        trace.getActiveSpan().addEvent("fetch " + url, {
          "response.body": resultText, // this gives me {} all the time and I don't know why
          "http.url": url,
          "http.status": result.status,
        });
        console.log("Response from " + url + ": " + resultText);
      }

      try {
        return JSON5.parse(resultText);
      } catch (e) {
        return {
          fetchEventError: true,
          statusCode: result.status,
          message: `Response to ${url} was not parsable as JSON: ` + resultText,
        };
      }
    },
    (error) => {
      recordError(error, { "http.url": url });
      return { fetchEventError: true, statusCode: 0, message: error.message };
    }
  );
}
