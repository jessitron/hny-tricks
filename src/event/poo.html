<div class="event-selection">
  <p>Choose an event to send:</p>
  <div>
    <input
      type="radio"
      id="event1"
      name="event_choice"
      value="event1"
      checked="true"
    /><label for="event1">one root span</label
    ><a
      href="https://gist.github.com/jessitron/c1e01f2d6aebde4d5ee804aefa18ac7d"
      target="_blank"
      ><img src="external-link.svg" class="icon" alt="view event definition" /></a>
  </div>
  <div>
    <input type="radio" id="event2" name="event_choice" value="event2" /><label
      for="event2"
      >structured log</label
    ><a
      href="https://gist.github.com/jessitron/ec40d3622721177d2fbdb7195aac2c39"
      target="_blank"
      ><img src="external-link.svg" class="icon" alt="view event definition" /></a>
  </div>
  <div>
    <input
      type="radio"
      id="custom_event"
      name="event_choice"
      value="custom"
    /><label for="custom_event">custom:</label
    ><input
      type="text"
      name="custom_url"
      placeholder="Enter raw JSON URL"
      style="margin-left: 1em; width: 20em"
      oninput="if(this.value) document.querySelector('#custom_event').checked = true"
    />
  </div>
</div>
