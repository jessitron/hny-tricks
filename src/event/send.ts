import { trace } from "@opentelemetry/api";
import { HnyTricksAuthorization, constructEnvironmentLink } from "../common";
import { DatasetSlug } from "../datasets/dataset_common";
import { fetchFromHoneycombApi, isFetchError } from "../HoneycombApi";
import { html } from "../htm-but-right";
import { StatusUpdate } from "../status";
import { fetchEventJson, isFetchEventError } from "./fetch";
import { RandomIdGenerator } from "./RandomIdGenerator";

type AvailableEvent = {
  description: string;
  userViewableUrl: string;
  rawUrl: string;
};

export const AVAILABLE_EVENTS: Record<string, AvailableEvent> = {
  event1: {
    description: "one root span",
    userViewableUrl:
      "https://gist.github.com/jessitron/c1e01f2d6aebde4d5ee804aefa18ac7d",
    rawUrl:
      "https://gist.github.com/jessitron/c1e01f2d6aebde4d5ee804aefa18ac7d/raw",
  },
  event2: {
    description: "structured log",
    userViewableUrl:
      "https://gist.github.com/jessitron/ec40d3622721177d2fbdb7195aac2c39",
    // for some reason, the raw URL for this one doesn't update when I edit the gist :-(
    rawUrl:
      "https://gist.githubusercontent.com/jessitron/ec40d3622721177d2fbdb7195aac2c39/raw/0f1bba813f1454dd374f9db3344bc6b732dcc1b9/structured-log.json",
  },
};

function queryByTransmission(transmissionId: string) {
  return {
    time_range: 7200,
    granularity: 0,
    breakdowns: [],
    calculations: [
      {
        op: "COUNT",
      },
    ],
    filters: [
      {
        column: "hny-tricks.transmission_id",
        op: "=",
        value: transmissionId,
      },
    ],
    filter_combination: "AND",
    orders: [],
    havings: [],
    trace_joins: [],
    limit: 100,
  };
}

/* here is how to get a slug from a name, from the Slugify function in hound.
 * (exceptions - if it needs disambiguation, that'd be random. It also can't be over 175 chars)
 * slug := slugReplaceRegex.ReplaceAllString(strings.ToLower(name), "-")
 */
const slugReplaceRegex = /[^a-z0-9_~\.-]/g;
function slugify(datasetName: string): DatasetSlug {
  return datasetName.replaceAll(slugReplaceRegex, "-");
}

export type SendEventInput = {
  service_name: string;
  event_choice: "custom" | keyof typeof AVAILABLE_EVENTS;
  custom_url?: string;
};

export async function sendEvent(
  auth: HnyTricksAuthorization,
  input: SendEventInput
): Promise<StatusUpdate> {
  const gen = new RandomIdGenerator();
  const traceId = gen.generateTraceId();
  const spanId = gen.generateSpanId();
  const transmissionId = gen.generateSpanId();
  const datasetName = input.service_name;
  const span = trace.getActiveSpan();

  const rawUrl =
    input.event_choice === "custom"
      ? input.custom_url
      : AVAILABLE_EVENTS[input.event_choice]?.rawUrl;
  if (!rawUrl) {
    return {
      success: false,
      html: "I need a URL to pull some JSON from",
    };
  }

  const rawData = await fetchEventJson(rawUrl);
  if (isFetchEventError(rawData)) {
    return { success: false, html: rawData.message };
  }

  // minimal change. Only 1 URL expected, it contains an object.
  const data = {
    ...rawData,
    "trace.trace_id": traceId,
    "trace.span_id": spanId,
    "hny-tricks.transmission_id": transmissionId,
    service_name: datasetName,
  };

  const datasetSlug = slugify(input.service_name);
  span.setAttributes({
    "app.send.datasetName": input.service_name,
    "app.send.datasetSlug": datasetSlug,
  });
  const url = "events/" + encodeURIComponent(datasetName);

  const result = await fetchFromHoneycombApi(
    {
      apiKey: auth.apiKey,
      keyInfo: auth.keyInfo,
      method: "POST",
      body: data,
      whatToExpectBack: "string",
    },
    url
  );
  if (isFetchError(result)) {
    return { success: false, html: result.message };
  }

  const linkToQuery =
    constructEnvironmentLink(auth) +
    "datasets/" +
    datasetSlug +
    "?query=" +
    encodeURIComponent(JSON.stringify(queryByTransmission(transmissionId))) +
    "&tab=explore";

  return {
    success: true,
    html: html`<span>
      1 event sent. <a href=${linkToQuery} target="_blank"> ðŸ“‰ </a>
    </span>`,
  };
}
