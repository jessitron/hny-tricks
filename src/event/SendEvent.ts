import { trace } from "@opentelemetry/api";
import { HnyTricksAuthorization } from "../common";
import { Html, html } from "../htm-but-right";
import { AVAILABLE_EVENTS, AvailableEvent } from "./send";

export function sendEventSection(auth: HnyTricksAuthorization, status?: Html) {
  const span = trace.getActiveSpan();
  span?.setAttributes({
    "app.datasets.canSendEvents": auth.permissions.canSendEvents,
  });
  if (!auth.permissions.canSendEvents) {
    return html`<section id="send-event-section">
      <div class="section-unavailable">
        <h3 class="section-title unavailable">Send a test span</h3>
        <p class="unavailable">
          Unavailable. This API key does not have permissions to send events.
        </p>
      </div>
    </section>`;
  }

  return html`<section id="send-event-section">
    <h3 class="section-title">Send a test span</h3>
    <form>
      <div class="event-form">
        <${EventSelection} availableEvents=${AVAILABLE_EVENTS} />
        <div>
          <label for="service_name">Service name (determines dataset):</label>
          <input name="service_name" value="testy-mctesterson" />
          <button
            hx-post="/event/send"
            hx-include="#auth_data"
            hx-target="#send-event-section"
            hx-swap="outerHTML"
          >
            Send
          </button>
        </div>
      </div>
    </form>
    ${status}
  </section>`;
}

export function justThisPart() {
  return html`<body><${EventSelection} availableEvents=${AVAILABLE_EVENTS}></body>`;
}

export function EventSelection(params: {
  availableEvents: Record<string, AvailableEvent>;
}) {
  return html`<div class="event-selection"></div>`;
}

export function presetEventsBit(params: {
  availableEvents: Record<string, AvailableEvent>;
}) {
  return html`<><p>Choose an event to send:</p>
    ${Object.entries(params.availableEvents)
      .map(([key, event], index) => html`<span>${key}</span>`)
      .join("")} </>`;
}

export function onePresetEvent(key: string, event: AvailableEvent) {
  return html`<div>
    <input
      type="radio"
      id=${key}
      name="event_choice"
      value=${key}
      checked=${key === "event1"}
    />
    <label for=${key}>${event.description}</label>
    <a href=${event.userViewableUrl} target="_blank">
      <img src="external-link.svg" class="icon" alt="view event definition" />
    </a>
  </div>`;
}

export function customEventBit() {
  return html`<div>
    <input type="radio" id="custom_event" name="event_choice" value="custom" />
    <label for="custom_event">custom:</label>
    <input
      type="text"
      name="custom_url"
      placeholder="Enter raw JSON URL"
      style="margin-left: 1em; width: 20em;"
      oninput="if(this.value) document.querySelector('#custom_event').checked = true"
    />
  </div>`;
}
