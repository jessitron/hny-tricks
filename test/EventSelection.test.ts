import { AVAILABLE_EVENTS } from "../src/event/send";
import {
  customEventBit,
  EventSelection,
  onePresetEvent,
  presetEventsBit,
} from "../src/event/SendEvent";

describe("EventSelection", () => {
  it("renders the event selection form", () => {
    const result = EventSelection({ availableEvents: AVAILABLE_EVENTS });
    expect(result).not.toContain("Invalid HTML detected");
    expect(result).toMatchSnapshot();
  });
});

describe("customEventBit", () => {
  it("renders", () => {
    const result = customEventBit();
    expect(result).not.toContain("Invalid HTML detected");
    expect(result).toMatchSnapshot();
  });
});

describe("presetEventsBit", () => {
  it("renders", () => {
    const result = presetEventsBit({ availableEvents: AVAILABLE_EVENTS });
    expect(result).not.toContain("Invalid HTML detected");
    expect(result).toMatchSnapshot();
  });
});

describe("onePresetEvent", () => {
  it("renders", () => {
    const result = onePresetEvent("event1", AVAILABLE_EVENTS.event1);
    expect(result).not.toContain("Invalid HTML detected");
    expect(result).toMatchSnapshot();
  });
});
