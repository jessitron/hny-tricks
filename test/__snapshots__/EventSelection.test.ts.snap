// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventSelection renders the event selection form 1`] = `"<div class="event-selection"></div>"`;

exports[`customEventBit renders 1`] = `"<div><input type="radio" id="custom_event" name="event_choice" value="custom"><label for="custom_event">custom:</label><input type="text" name="custom_url" placeholder="Enter raw JSON URL" style="margin-left: 1em; width: 20em;" oninput="if(this.value) document.querySelector(&apos;#custom_event&apos;).checked = true"></div>"`;

exports[`onePresetEvent renders 1`] = `"<div><input type="radio" id="event1" name="event_choice" value="event1" checked="true"><label for="event1">one root span</label><a href="https://gist.github.com/jessitron/c1e01f2d6aebde4d5ee804aefa18ac7d" target="_blank"><img src="external-link.svg" class="icon" alt="view event definition"></a></div>"`;

exports[`presetEventsBit renders 1`] = `"<p>Choose an event to send:</p>&lt;span&gt;event1&lt;/span&gt;&lt;span&gt;event2&lt;/span&gt; "`;
